### PR TITLE
fix: audit issue 9

### PR DESCRIPTION
**Pull request by gitmp01**
Previously using a token with different decimals was introducing conversion errors during the withdraw and deposit of funds from/to the Lockbox contract.

Now a conversion ratio is applied during this operation.

**Note:** the pNetwork protocol assumes that the created XERC20 tokens will always have 18 decimals.

**Note:** user may see his balance reduced for dust withdrawals, frontend is in charge to avoid that edge case

### CODE REVIEW

File: solidity/test/forge/XERC20Lockbox.t.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                emptyBridges,
+                freezingEnabled
+            )
+        );
+
+        lockbox = XERC20Lockbox(
+            factory.deployLockbox(address(xerc20), address(USDC), isNative)
+        );
+
+        deal(address(USDC), user, usdcBalance);
```
Comment (envin3) [inline solidity/test/forge/XERC20Lockbox.t.sol:73]: 👍

File: solidity/src/contracts/XERC20Lockbox.sol [hunk 2]
```diff
@@ -125,7 +138,8 @@ contract XERC20Lockbox is IXERC20Lockbox {
       (bool _success,) = payable(_to).call{value: _amount}('');
       if (!_success) revert IXERC20Lockbox_WithdrawFailed();
     } else {
-      ERC20.safeTransfer(_to, _amount);
+      uint256 amount = _amount / conversionRatio;
```
Comment (envin3) [inline solidity/src/contracts/XERC20Lockbox.sol:141]: Ok

File: solidity/src/contracts/XERC20Lockbox.sol [hunk 3]
```diff
@@ -141,8 +155,19 @@ contract XERC20Lockbox is IXERC20Lockbox {
       ERC20.safeTransferFrom(msg.sender, address(this), _amount);
     }
 
-    XERC20.mint(_to, _amount);
-    emit Deposit(_to, _amount);
+    uint256 amount = _amount * conversionRatio;
```
Comment (envin3) [inline solidity/src/contracts/XERC20Lockbox.sol:158]: Ok

File: solidity/test/forge/XERC20Lockbox.t.sol [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    function test_userMustHaveExpectedInitialBalance() public onlyMainnet {
+        assertEq(USDC.balanceOf(user), usdcBalance);
+    }
+
+    function test_depositPrecision() public onlyMainnet {
+        vm.startPrank(user);
+        USDC.approve(address(lockbox), usdcBalance);
+        lockbox.deposit(usdcBalance);
+
+        assertEq(xerc20.balanceOf(user), usdcBalance * conversionRatio);
```
Comment (envin3) [inline solidity/test/forge/XERC20Lockbox.t.sol:94]: Ok

File: solidity/test/forge/XERC20Lockbox.t.sol [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        uint256 xerc20Balance = xerc20.balanceOf(user);
+        uint256 dust = 1000;
+        uint256 snapId = vm.snapshotState();
+
+        xerc20.approve(address(lockbox), dust);
+        lockbox.withdraw(dust);
+
+        assertEq(USDC.balanceOf(user), 0);
+        assertEq(USDC.balanceOf(address(lockbox)), usdcBalance);
+        assertEq(xerc20.balanceOf(user), xerc20Balance - dust);
```
Comment (envin3) [inline solidity/test/forge/XERC20Lockbox.t.sol:113]: Xerc20balance == erc20balance * conversionRatio checked on previous test, ok

File: solidity/test/forge/XERC20Lockbox.t.sol [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        assertEq(xerc20.balanceOf(address(lockbox)), 0);
+
+        vm.revertToState(snapId);
```
Comment (envin3) [inline solidity/test/forge/XERC20Lockbox.t.sol:116]: Nice

**Review by envin3 — APPROVED** [review]
